pkg/storage/couchdb/couchdbstore.go

```diff
@@ -196,16 +197,111 @@ func (c *CouchDBStore) Get(k string) ([]byte, error) {
 		return nil, err
 	}
 
-	_, containsAttachment := destinationData["_attachments"]
+	return c.getStoredValueFromRawDoc(rawDoc, k)
+}
+
+// CreateIndex creates an index based on the provided CreateIndexRequest.
+// createIndexRequest.IndexStorageLocation refers to the design doc that the index should get placed in.
+// createIndexRequest.IndexName is the name for the index that will be stored in CouchDB.
+// createIndexRequest.WhatToIndex must be a valid index object as defined here:
+//   http://docs.couchdb.org/en/stable/api/database/find.html#db-index
+func (c *CouchDBStore) CreateIndex(createIndexRequest storage.CreateIndexRequest) error {
+	return c.db.CreateIndex(context.Background(), createIndexRequest.IndexStorageLocation,
+		createIndexRequest.IndexName, createIndexRequest.WhatToIndex)
+}
+
+// Query executes a query using the CouchDB _find endpoint.
+func (c *CouchDBStore) Query(findQuery string) (storage.ResultsIterator, error) {
+	resultRows, err := c.db.Find(context.Background(), findQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	return &couchDBResultsIterator{store: c, resultRows: resultRows}, nil
+}
+
+type couchDBResultsIterator struct {
+	store      *CouchDBStore
+	resultRows *kivik.Rows
+	firstErr   error
+}
+
+// Next moves the pointer to the next value in the iterator. It returns false if the iterator is exhausted.
+// Note that the Kivik library automatically closes the kivik.Rows iterator if the iterator is exhaused.
+func (i *couchDBResultsIterator) Next() bool {
+	return i.resultRows.Next()
+}
+
+// Release releases associated resources. Release should always result in success
+// and can be called multiple times without causing an error.
+func (i *couchDBResultsIterator) Release() {
+	closeErr := i.resultRows.Close()
+
+	if i.firstErr == nil {
+		i.firstErr = closeErr
+	}
+}
+
+// Error returns the first error that occurred during iteration.
+// Exhausting all the key-value pairs is not considered to be an error.
+func (i *couchDBResultsIterator) Error() error {
+	if i.firstErr != nil {
+		return i.firstErr
+	}
+
+	return i.resultRows.Err()
+}
+
+// Warning returns the first warning that occurred during iteration.
+// Exhausting all the key-value pairs is not considered to be a warning.
+// Kivik only guarantees that this value will be set after all the rows have been iterated through.
+func (i *couchDBResultsIterator) Warning() string {
```

Comment:
Not sure if this is useful to the client. Normally if a warning occurs we just log it.

Reply:
This is the only way I can tell if a query I'm attempting is not using the index I specified (see the unit tests for an example). I thought this was important since CouchDB will still allow me to query using an improper index, but of course it'll just be slower (potentially much slower in the case of millions of documents?). I even think this should really be a full-blown error so that the client won't accidentally try to use an improper index. (for instance, me attempting to index an element in an array - had I not looked at the warning message, I wouldn't have known that my index was totally useless).

My preferred strategy would be to make the query() method throw an error if I see that "index not used" message in the kivk Warning() method. What do you think?

Reply:
I think that a log is good enough. The client may not call this function and we'd never know that a warning occurred.

Reply:
Fair point - OK, I'll make those changes